File: VidoX_Google/app/src/main/java/com/ashbash/vidoxproject/services/extraction/VideoExtracting.kt
package com.ashbash.vidoxproject.services.extraction

import com.ashbash.vidoxproject.models.VideoMetadata
import com.ashbash.vidoxproject.models.VideoPlatform
import java.net.URL

sealed class ExtractionError(message: String) : Exception(message) {
    data object InvalidURL : ExtractionError(
        "That doesn’t look like a valid link. Try pasting a full https:// URL."
    )
    data class UnsupportedPlatform(val platform: VideoPlatform) : ExtractionError(
        "Downloading from ${platform.displayName} isn’t available yet."
    )
    data object NotDirectMedia : ExtractionError(
        "This isn’t a direct media file. Paste a file link (.mp4, .mov, …) or a supported video page when standalone downloads are enabled."
    )
    data class Network(val detail: String) : ExtractionError(detail)
    data object EmptyResponse : ExtractionError("The server returned an empty response.")
}

sealed class PageExtractionError(message: String) : Exception(message) {
    data object Unsupported : PageExtractionError(
        "This link isn’t supported for download yet."
    )
    data object NoMediaFound : PageExtractionError(
        "Couldn’t find a downloadable video on that page. Try another link or a direct .mp4 URL."
    )
    data class Network(val detail: String) : PageExtractionError(detail)
    data object InvalidResponse : PageExtractionError(
        "The site returned data VidoX couldn’t read."
    )
}

fun interface VideoExtracting {
    suspend fun extract(from: URL): VideoMetadata
}

File: VidoX_Apple/VidoXProject/Services/Extraction/VideoExtracting.swift
import Foundation

/// Errors raised while resolving a URL into downloadable metadata.
enum ExtractionError: LocalizedError, Sendable {
    case invalidURL
    case unsupportedPlatform(VideoPlatform)
    case notDirectMedia
    case network(String)
    case emptyResponse

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            "That doesn’t look like a valid link. Try pasting a full https:// URL."
        case .unsupportedPlatform(let platform):
            "Downloading from \(platform.displayName) isn’t available yet."
        case .notDirectMedia:
            "This isn’t a direct media file. Paste a file link (.mp4, .mov, …) or a supported video page (YouTube, Vimeo, and similar) when standalone downloads are enabled."
        case .network(let message):
            message
        case .emptyResponse:
            "The server returned an empty response."
        }
    }
}

/// Turns a pasted URL into `VideoMetadata` with one or more formats.
protocol VideoExtracting: Sendable {
    func extract(from url: URL) async throws -> VideoMetadata
}
